Scan more than one component at the time

vivodikj opened the pull request:
**Summary**

This PR create an option for add more packages in scanning component feature, [inspired by this issue](https://github.com/InsertKoinIO/koin-annotations/issues/180). With this change, developers can use new `@ComponetsScan` annotation with `values` parameter where they can add all needed packages. The original `@ComponentScan` annotations is 100% preserved and now can cooperate with new `@ComponetsScan`.

All tests still passes like a charm and new test passed without any problem.

**Examples**

*Before*
```kotlin

@Module(includes = [DataModule::class])
@ComponentScan("org.koin.sample.androidx.app")
class AppModule

@Module(includes = [CommonModule::class, RepositoryModule::class])
@ComponentScan("org.koin.sample.androidx.data")
internal class DataModule

```

*After 1. variant*
```kotlin

@Module(includes = [CommonModule::class, RepositoryModule::class])
@ComponentsScan(values = ["org.koin.sample.androidx.app", "org.koin.sample.androidx.data"])
class AppModule

```

*After 2. variant*
```kotlin

@Module
@ComponentsScan(values = [
    "org.koin.sample.androidx.app", "org.koin.sample.androidx.data",
    "org.koin.sample.androidx.repository", "org.koin.sample.android.library",
])
class AppModule

```

*After 1. variant if we want scan actual package*
```kotlin

@Module
@ComponentsScan(values = [
    "", "org.koin.sample.androidx.app", "org.koin.sample.androidx.data",
    "org.koin.sample.androidx.repository", "org.koin.sample.android.library",
])
class AppModule

```

*After 2. variant if we want scan actual package*
```kotlin

@Module
@ComponentsScan
@ComponentsScan(values = [
    "org.koin.sample.androidx.app", "org.koin.sample.androidx.data",
    "org.koin.sample.androidx.repository", "org.koin.sample.android.library",
])
class AppModule

```

**NOTE:**

If dev use `@ComponentsScan` or `@ComponentsScan([])` or `@ComponentsScan([""])` this will scan actual package like original `@ComponentScan`.

If dev add non existing or correct package name to `@ComponentsScan` values, the package will be ignored and nothing happens.

If dev add multiple package names to `@ComponentsScan` values, only one will be used and multiple package names error is not shown.

arnaudgiuliani commented:
let's go on branch v2 as it's a new feature / improvement proposal

arnaudgiuliani commented:
I propose to keep one dedicated annotation, with `@ComponentScan`. Let's use the `values` atrtibutes to gather array of package.

No doubling of `@ComponentScan` needed. If people wants to go with current package, let's keep it "".

Can you update your code? 🙏

vivodikj commented:
@arnaudgiuliani 
I have modified the code, including the examples. The user can use `@ComponentScan` without additional parameters or "" to scan the current package. I also added a trim to each scanned package, so there shouldn't be a problem with, for example "         ".

arnaudgiuliani commented on projects/koin-annotations/src/commonMain/kotlin/org/koin/core/annotation/CoreAnnotations.kt:
```diff
@@ -202,13 +202,14 @@ annotation class PropertyValue(val value: String)
  */
 @Target(AnnotationTarget.CLASS, AnnotationTarget.FIELD)
-annotation class ComponentScan(val value: String = "")
+annotation class ComponentScan(val values: Array<String> = [""])
```
could be empty array here?

arnaudgiuliani replied:
or perhaps useless 🤔

vivodikj replied:
oki, I'll remove default empty string value.

arnaudgiuliani commented:
CI failed. Can you look at this? @vivodikj :

```
e: file:///Users/runner/work/koin-annotations/koin-annotations/examples/android-coffee-maker/build/generated/ksp/debug/kotlin/org/koin/ksp/generated/KoinDefault-563727094.kt:32:2 Unresolved reference 'defineScopedStuff'.
> Task :android-coffee-maker:compileDebugKotlin FAILED
e: file:///Users/runner/work/koin-annotations/koin-annotations/examples/android-coffee-maker/src/main/java/org/koin/sample/androidx/di/AppModules.kt:18:16 No parameter with name 'values' found.
```

vivodikj commented:
@arnaudgiuliani about the CI failed. It's look like CI don't use new "updated" version of coin annotations. Maybe they don't use created mavenLocal repository but mavenCentral repository. Where is not new componentScan annotation with values array.

arnaudgiuliani commented:
@vivodikj I'm about to merge your PR. We have a conflict in the sample project

vivodikj commented:
@arnaudgiuliani  No more conflict :)

arnaudgiuliani commented:
Great job 👍

arnaudgiuliani commented:
I'm sorry. I'm gathering all the release stuff. The problem is that we are breaking compatibility on @ComponentScan. This was also your proposal to add another annotation. :/ 

Let's find how too mitigate that. We could use an extra Annotation property? 🤔 
This PR is also bringing capacity on Scanning: https://github.com/InsertKoinIO/koin-annotations/pull/131

vivodikj commented:
Yes, that's why I created another annotation in the first place.

We could add extra new property to existing `@componentScan` annotation, something like nullable string array values? Or we can create new annotation as before. 

Ya, I think this PR with #131 will be very good combination.

vivodikj commented:
@arnaudgiuliani Maybe I found a pretty good solution. We can make value in `@ComponentScan` vararg. Which will look like `annotation class ComponentScan(vararg val value: String = [])` and usage can be for for example `@ComponentScan("....", ".....", ".....")`, `@ComponentScan("....")`, `@ComponentScan` also `@ComponentScan(value = ["...", "..."])` or `@ComponentScan(value = ["..."])`. But there don't work named parameter this way `@ComponentScan(value = "...")` only this `@ComponentScan(value = ["..."]) ` or `@ComponentScan(".....")`.

arnaudgiuliani commented:
Yeah, clearly. Let's add the `paths` attribute like this:

```kotlin
/**
 * Gather definitions declared with Koin definition annotation
 * Will scan in the current package or an explicit package path name
 * Also can scan across multiple package paths. 
 *
 * @param value: path to scan
 * @param paths : list of path to scan
 */
@Target(AnnotationTarget.CLASS, AnnotationTarget.FIELD)
annotation class ComponentScan(val value: String = "", val paths: Array<String> = [])
```

What do you think?

arnaudgiuliani commented:
> @arnaudgiuliani Maybe I found a pretty good solution. We can make value in `@ComponentScan` vararg. Which will look like `annotation class ComponentScan(vararg val value: String = [])` and usage can be for for example `@ComponentScan("....", ".....", ".....")`, `@ComponentScan("....")`, `@ComponentScan` also `@ComponentScan(value = ["...", "..."])` or `@ComponentScan(value = ["..."])`. But there don't work named parameter this way `@ComponentScan(value = "...")` only this `@ComponentScan(value = ["..."]) ` or `@ComponentScan(".....")`.

our messages get crossed. Yeah interesting proposal. Not sure people used @ComponentScan with `value` property 🤔

vivodikj commented:
Me and my team don't use `@ComponentScan` with named `value` property but only `@ComponentScan("...")`. But if someone use named value property, they can delete it and everything will works. I think this is little bit cleaner to use, but if you like your proposal more I will make it. Its up to you.

arnaudgiuliani commented:
Question here, why your code is hosted on 1.4.0 branch?

vivodikj commented:
@arnaudgiuliani Ok, I'll rebase it.

Because I started developing it on version 1.4. And version 2.0 wasn't out yet.

arnaudgiuliani commented:
Ok. We are waiting for your last PR to go 👍

vivodikj commented:
@arnaudgiuliani No more conflict. I apologize for the long wait, but I had a lot of work.

arnaudgiuliani commented:
No worry. I'll take a look
